plugin/pkg/shoot/validator/admission.go

```diff
@@ -519,9 +519,16 @@ func validateKubernetesVersionConstraints(constraints []core.ExpirableVersion, s
 		}
 
 		if getLatestPatchVersion {
+			// ignore preview versions for defaulting
+			if versionConstraint.Classification != nil && *versionConstraint.Classification == core.ClassificationPreview {
+				// do not include preview version in the valid values when using defaulting
+				validValues = validValues[:len(validValues)-1]
```

Comment:
It can only be, it has been [previously added here](https://github.com/gardener/gardener/blob/70b7ea5c9831489463276478936d19295795d8d2/plugin/pkg/shoot/validator/admission.go#L515)

Reply:
Ah, sure. My bad... 

Still, I think, appending an item to a slice, then removing it directly afterwards if certain conditions are met, is not good style.
WDYT about doing it like this?
```go
if !getLatestPatchVersion {
  validValues = append(validValues, versionConstraint.Version)
} else {
  // ignore preview versions for defaulting
  if versionConstraint.Classification == nil || *versionConstraint.Classification != core.ClassificationPreview {
    validValues = append(validValues, versionConstraint.Version)
  }
  // and so on...
}
```

Reply:
I'm not quite sure how this is related to the `getLatestPatchVersion` branch. Why not checking the `preview` classification when the version is added to the `validValues` in line 515?

Reply:
@timuthy A preview version is a valid version. A shoot can be created with it. Otherwise no shoot can use a preview version.

Reply:
Adjusted it according to @tim-ebert comment